src/redjubjub.rs

```diff
@@ -0,0 +1,133 @@
+//! Implementation of RedJubjub, a specialization of RedDSA to the Jubjub curve.
+//! See section 5.4.6 of the Sapling protocol specification.
+
+use pairing::Field;
+use rand::Rng;
+
+use jubjub::{FixedGenerators, JubjubEngine, JubjubParams, Unknown, edwards::Point};
+use util::hash_to_scalar;
+
+fn h_star<E: JubjubEngine>(a: &[u8], b: &[u8]) -> E::Fs {
+    hash_to_scalar::<E>(b"Zcash_RedJubjubH", a, b)
+}
+
+pub struct Signature<E: JubjubEngine> {
+    r: Point<E, Unknown>,
+    s: E::Fs,
+}
+
+pub struct PrivateKey<E: JubjubEngine>(E::Fs);
+
+pub struct PublicKey<E: JubjubEngine>(Point<E, Unknown>);
+
+impl<E: JubjubEngine> PrivateKey<E> {
+    pub fn randomize(&self, alpha: E::Fs) -> Self {
+        let mut tmp = self.0;
+        tmp.add_assign(&alpha);
+        PrivateKey(tmp)
+    }
+
+    pub fn sign<R: Rng>(&self, msg: &[u8], rng: &mut R, params: &E::Params) -> Signature<E> {
+        // T = (l_H + 128) bits of randomness
+        // For H*, l_H = 512 bits
+        let mut t = [0u8; 80];
+        rng.fill_bytes(&mut t[..]);
+
+        // r = H*(T || M)
+        let r = h_star::<E>(&t[..], msg);
+
+        // R = r . G
+        let g_r = params
+            .generator(FixedGenerators::SpendingKeyGenerator)
+            .mul(r, params);
+        let mut rbar = [0u8; 32];
+        g_r.write(&mut rbar[..])
+            .expect("Jubjub points should serialize to 32 bytes");
+
+        // S = r + H*(Rbar || M) . sk
+        let mut s = h_star::<E>(&rbar[..], msg);
+        s.mul_assign(&self.0);
+        s.add_assign(&r);
+
+        Signature { r: g_r.into(), s }
+    }
+}
+
+impl<E: JubjubEngine> PublicKey<E> {
+    pub fn from_private(privkey: &PrivateKey<E>, params: &E::Params) -> Self {
+        let res = params
+            .generator(FixedGenerators::SpendingKeyGenerator)
+            .mul(privkey.0, params)
+            .into();
+        PublicKey(res)
+    }
+
+    pub fn randomize(&self, alpha: E::Fs, params: &E::Params) -> Self {
+        let res: Point<E, Unknown> = params
+            .generator(FixedGenerators::SpendingKeyGenerator)
+            .mul(alpha, params)
+            .into();
+        let res = res.add(&self.0, params);
+        PublicKey(res)
+    }
+
+    pub fn verify(&self, msg: &[u8], sig: &Signature<E>, params: &E::Params) -> bool {
+        // c = H*(Rbar || M)
+        let mut rbar = [0u8; 32];
+        sig.r
+            .write(&mut rbar[..])
+            .expect("Jubjub points should serialize to 32 bytes");
+        let c = h_star::<E>(&rbar[..], msg);
+
+        // S . G = R + c . vk
```

Comment:
Do we want to follow the RFC for EdDSA? In that case we need to check for the canonical signature representation, i.e. that 0 <= S < order of the curve.

Reply:
This is in the spec definition of RedDSA.Verify already, but typo'd - we should ensure S < $\ParamJ{r}

Reply:
Additional comment because I forget to start a review